### PR TITLE
NHSNLINK-384: Create Practitioner resources for each user that makes …

### PR DESCRIPTION
**Pull request by arianamihailescu**
…a request

### CODE REVIEW

File: api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        this.config = config;
+    }
+
+
+    @Override
+    public void onAuthenticationSuccess(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, Authentication authentication) throws IOException, ServletException {
+        IGenericClient client = FhirContext.forR4().newRestfulGenericClient(config.getFhirServerStore());
+        Practitioner practitioner = ((LinkCredentials) authentication.getPrincipal()).getPractitioner();
+        try {
+            client.read().resource(Practitioner.class).withId("2").execute();
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java:33]: `.withId("2")` is incorrectly hard-coded, here, I believe.
I think it should be `.withId(practitioner.getIdElement().getIdPart())`?

File: api/src/main/java/com/lantanagroup/link/api/ApiSecurityConfig.java [hunk 2]
```diff
@@ -22,14 +31,18 @@
...
   @Autowired
   private Environment env;
 
   @Override
   protected void configure(HttpSecurity http) throws Exception {
-    PreAuthTokenHeaderFilter authFilter = new PreAuthTokenHeaderFilter("Authorization");
+    PreAuthTokenHeaderFilter authFilter = new PreAuthTokenHeaderFilter("Authorization", linkCredentials);
     authFilter.setAuthenticationManager(new LinkAuthManager(this.config));
-
+    authFilter.setAuthenticationSuccessHandler(new LinkAuthenticationSuccessHandler(this.config));
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/ApiSecurityConfig.java:45]: Nice! Didn't know this existed. Good pattern...

File: api/src/main/java/com/lantanagroup/link/api/auth/PreAuthTokenHeaderFilter.java [hunk 3]
```diff
@@ -2,31 +2,78 @@
...
-
     return null;
   }
+
+  @Override
+  protected Object getPreAuthenticatedCredentials (HttpServletRequest request) {
+    return request.getHeader(authHeaderName);
+  }
+
+  private Practitioner getPractitioner (DecodedJWT jwt) {
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/auth/PreAuthTokenHeaderFilter.java:47]: Consider renaming to `toPractitioner()` and making it static. I don't think this relies on any `this` properties...

File: api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        } catch (ResourceNotFoundException ex) {
+            String msg = String.format("Practitioner Resource with identifier  \"%s\"  was not found on FHIR server \"%s\". It will be created.", practitioner.getId(), config.getFhirServerStore());
```
Comment (seanmcilvenna) [inline api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java:35]: Nice work on the log messages. This is the perfect level of detail for production debugging.

**Review by seanmcilvenna — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java (hunk 1 above)
Comment (arianamihailescu) [reply api/src/main/java/com/lantanagroup/link/api/auth/LinkAuthenticationSuccessHandler.java]: Corrected

**Review by seanmcilvenna — APPROVED** [review]

**Pull request merged**